Docs: Interactivity API : New pages - About and FAQ

juanmaguitar opened the pull request:
## What?

Creates two new pages to get a deeper understanding of what the Interactivity API is or find answers to questions about this standard:
- **About the Interactivity API** To learn more about the API Goals and the reasoning behind the use of a standard to add interactivity to blocks.
- **Frequently Asked Questions** To find responses to some frequently asked questions about the technology behind and alternatives.

## Why?

To surface some explanations "hidden" on the proposal that are really useful for new users of the Interactivity API to help them understand the reasoning behind this standard.

This PR solves https://github.com/WordPress/gutenberg/issues/60783 among other questions 

cc: @luisherranz @cbravobernal @SantosGuillamot

juanmaguitar committed:
new pages - about and faq

sirreal commented on docs/reference-guides/interactivity-api/iapi-faq.md:
```diff
@@ -0,0 +1,129 @@
+- It’s battle-tested.
+- It’s performant (even more when used with signals).
+- It’s compatible with React (through preact/compat, useful to share the same Editor components for some use cases where SSR is not important, and the components are very complex, like an e-commerce cart and checkout, for example).
+- It’s HTML-friendly (unlike React).
```
I'm not sure what this means, maybe this could link to context or be removed.

cbravobernal replied:
I agree on removing these ones.
 - We don't support components yet.
 - We should not talk about performance without metrics.
 - It's battle tested is not so important, as being in Core means already that.

juanmaguitar replied:
Thanks for the feedback.
I have removed these points 👍

sirreal commented on docs/reference-guides/interactivity-api/iapi-faq.md:
```diff
@@ -0,0 +1,129 @@
+
+## Do I need to know React, PHP, and this new Interactivity API?
+
+If you want to add frontend interactivity to your blocks using this API, the short answer is yes. If your block is not interactive, the block creation workflow will remain exactly the same.
+
+This API only adds a new standard way to easily add frontend interactivity to blocks. This means that you still need to use React to handle the editor part of your blocks.
```
The Interactivity API is not limited to blocks. That may be where it's most often used, but I'd like to avoid assumptions that it's only for blocks. I think it's _good_ to mention blocks, but can we also mention more general "interactive behaviors" or something general like that?

juanmaguitar replied:
I agree with expanding the Interactivity API's use case beyond blocks, but I would still prioritize the focus on blocks as the best way to get people started with the Interactivity API.

> That may be where it's most often used, but I'd like to avoid assumptions that it's only for blocks.

I would keep the mentions of blocks to ease the transition, but I would also introduce the idea of the Interactivity API's capacity to work beyond blocks. This could be another good FAQ 😊

Right now, there are no major use cases for the Interactivity API outside a block, and focusing on blocks helps developers better understand and start using it.

juanmaguitar replied:
@sirreal I have added the question, "Can the Interactivity API be used beyond a block?" to the FAQ. 
Please let me know your thoughts.

sirreal commented on docs/reference-guides/interactivity-api/iapi-faq.md:
```diff
@@ -0,0 +1,129 @@
+The API has been designed with performance in mind, so it shouldn’t be a problem:
+
+- **The runtime code needed for the directives is just ~10 KB**, and it only needs to be loaded once for all the blocks.
+- **It only loads the directives needed** by the blocks present on the page. For example, if no blocks use data-wp-show, that directive won’t be loaded.
```
I'm not sure, but I don't think this is accurate. @DAreRodz maybe you can confirm?

At any rate, directives are trivially small. The kind of benefit you'd get from skipping a single directive probably isn't worth it.

Maybe we just remove this bullet point.

cbravobernal replied:
I could not find directives being loaded conditionally too.

SantosGuillamot replied:
I think it was included in the original proposal post as an idea to improve performance if necessary. Additionally, it could become an "issue" if creating custom directives was allowed.

I agree that, as custom directives aren't enabled yet, and this improvement hasn't been implemented, we can remove the bullet point.

juanmaguitar replied:
Removed! 👍

sirreal commented on docs/reference-guides/interactivity-api/iapi-faq.md:
````diff
@@ -0,0 +1,129 @@
+
+No. Right now, directives only work in the frontend of blocks. However, it’ll be investigated whether some directives (and also your custom directives) could be reused across the frontend and the editor. It’s worth emphasizing that the realities of the editor application and the frontend of a website are very different, particularly around the interactivity they afford. It needs to be ensured tha ...
+
+## Does it work with the Core Translation API?
+
+It does! As the Interactivity API works perfectly with server-side rendering, you can use all the WordPress APIs including [`__()`](https://developer.wordpress.org/reference/functions/__/) and [`_e()`](https://developer.wordpress.org/reference/functions/_e/). You can use it to translate the text in the HTML (as you normally would) and even use it inside the store when using `wp_initial_state()` o ...
+
+```php
+// render.php
+wp_interactivity_state( 'favoriteMovies', array(
+      "1" => array(
+        "id" => "123-abc",
+        "movieName" => __("someMovieName", "textdomain")
+      ),
+) );
+```
````
This section seems misleading. Static strings in the interactive HTML can be rendered in PHP via the Translation API, but JavaScript translations aren't really supported yet in Script Modules. https://core.trac.wordpress.org/ticket/60234

cbravobernal replied:
We showcased the translation support in a WCEU. So I would keep it, but also explain how to set translatable JS strings and mention that would be improved.

juanmaguitar replied:
If I understand this properly, at this stage, translations within the Interactivity API are available on the server functions (PHP) but **still** not on the client methods (Javascript). 

I have adjusted the explanation and added the following at the end of the response:

> A translation API compatible with script modules (needed for the Interactivity API) is 
currently being worked on. Check https://core.trac.wordpress.org/ticket/60234 to follow the progress on this work.

luisherranz replied:
For non-interactive translated texts, you can just add them in PHP.

```php
<span><?php _e('hello'); ?></span>
```

For interactive translated texts, you can add the strings of translated text to the state in PHP and then use them on the client side:

```php
wp_interactivity_state('...', array(
  'hello' => __('hello'),
  'bye' => __('bye'),
  'welcomeText' => __('hello') // initial value of derived state for SSR
));
```

```html
<span data-wp-text="state.welcomeText"></span>
```

```js
store("...", {
  state: {
    get welcomeText() {
      return state.isUserLeaving ? state.bye : state.hello;
    },
  },
});
```

What is harder to do is things like plurals, numbers and such...

sirreal commented on docs/reference-guides/interactivity-api/iapi-about.md:
````diff
@@ -0,0 +1,191 @@
+_Dynamic block example_
+```html
+<div
+  data-wp-interactive='{ "namespace": "wpmovies" }'
````
We can use the string form here:

```suggestion
  data-wp-interactive="wpmovies"
```

juanmaguitar replied:
Fixed ✅

sirreal commented on docs/reference-guides/interactivity-api/iapi-about.md:
````diff
@@ -0,0 +1,191 @@
+```html
+<div
+  data-wp-interactive='{ "namespace": "wpmovies" }'
+  data-wp-context='{ "isOpen": false }'
````
This should use `wp_interactivity_data_wp_context`:

```suggestion
  <?php echo wp_interactivity_data_wp_context( array( 'isOpen' => false ) ); ?>
```

sirreal replied:
I think this suggestion was missed, let's always use the `wp_interactivity_data_wp_context` function, it's a best practice folks should follow.
```suggestion
  <?php echo wp_interactivity_data_wp_context( array( 'isOpen' => false ) ); ?>
```

sirreal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+<!-- Render.php file -->
+ 
+<div
+  data-wp-interactive='{ "namespace": "wpmovies" }'
+  data-wp-context="{ 'isOpen': true }"
```
Same - string-only namespace and use the `wp_interactivity_data_wp_context` function.

sirreal commented on docs/reference-guides/interactivity-api/iapi-about.md:
````diff
@@ -0,0 +1,191 @@
+</div>
+```
+
+Using imperative code may be easier when creating simple user experiences, but it becomes much more difficult as blocks become more complex. The Interactivity API must cover all use cases, from the simplest to the most challenging. That’s why a declarative approach using directives better fits the Interactivity API.
````
This is another chance to avoid being so block-centric:

```suggestion
Using imperative code may be easier when creating simple user experiences, but it becomes much more difficult as applications become more complex. The Interactivity API must cover all use cases, from the simplest to the most challenging. That’s why a declarative approach using directives better fits the Interactivity API.
```

sirreal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+The API has been designed to be as performant as possible:
+
+- **The runtime code needed for the directives is just ~10 KB**, and it only needs to be loaded once for all the blocks.
+- **It only loads the directives needed** by the blocks present on the page. For example, if no blocks are using data-wp-show, the code for this directive won’t be loaded.
```
This is repeated from the other page and I don't _think_ it's accurate.

sirreal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+
+- **The runtime code needed for the directives is just ~10 KB**, and it only needs to be loaded once for all the blocks.
+- **It only loads the directives needed** by the blocks present on the page. For example, if no blocks are using data-wp-show, the code for this directive won’t be loaded.
+- **The scripts will load without blocking the page rendering**.
```
It might be good to always use "script modules"

sirreal reviewed (commented):
I've left a bunch of feedback, but I'll leave final reviews for the other folks you mentioned who are better informed than I am.

cbravobernal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+To address this challenge the following requirements/goals for the Interactivity API were defined:
+
+- **Block-first and PHP-first**: The API must work well with PHP and the current block system, including dynamic blocks, widely extended in WordPress. It must support server-side rendering. Server-rendered HTML and client-hydrated HTML must be exactly the same. This is important for SEO and the user experience.
+- **Backward compatible**: The API must be compatible with WordPress hooks, which could, for example, modify server-rendered HTML. It must also be compatible with internationalization and existing JS libraries on the site (such as jQuery).
```
It is compatible with any existing JS libraries, but it won't work out the box. 

If your block uses jQuery, it won't be compatible with client side navigation. Also, if you add any event with any external library, to any element that its visibility is handled by the Interactivity API, you need to control its behaviour by your self.

SantosGuillamot replied:
Compatibility with other blocks is mentioned in [the next line](https://github.com/WordPress/gutenberg/pull/61323/files#diff-fcd8a995f6703c8a130f5507361a3fb7ad50216febc9899718d0fd74c0eb9777R19). Maybe we can just remove that part from here.

Apart from that, I believe that client side navigation probably deserves a separate explanation at some point. It is a complex topic with some nuances.

juanmaguitar replied:
This section discusses the requirements/goals that were defined for the Interactivity API, so I think no further explanations are needed (as this is more of a wishlist). We could add links to other parts of the docs for more details and further explanations in other sections (more about specific implementations of these ideas).  


I think the idea of the Interactivity API's capacity to coexist with other libraries (such as jQuery) is a powerful one that should be highlighted as it would help with its "Optional and gradual adoption" as remarked in the following point.

> , I believe that client side navigation probably deserves a separate explanation at some point. It is a complex topic with some nuances.

I completely agree with this. This is already reflected in [these plans for the documentation 
](https://github.com/WordPress/gutenberg/issues/53296#issuecomment-1911915212), and a dedicated section in the Docs will be added once client-side navigation can be fully implemented with the iAPI. In the meantime I'll provide a link to https://github.com/WordPress/gutenberg/issues/60951 to point to the progress on this work.

cbravobernal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+
+### Block-first and PHP-friendly
+
+The API is designed for the world of blocks and takes WordPress history of being closely attached to web standards to heart.
```
I don't fully understand this sentence 😅

juanmaguitar replied:
Any suggestions for a better rephrasing?
cc: @SantosGuillamot @sirreal

sirreal replied:
I'm not really sure, maybe "designed to integrate seamlessly with blocks"? What's the part that's problematic, @cbravobernal? I do agree, it's hard to rephrase this because I'm not sure what exactly we're trying to express here. Should we draw a parallel where blocks are HTML markups + HTML comments, and directives are _just_ HTML attributes?

I'm not sure what PHP-friendly means, is this about server-side rendering?

cbravobernal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+
+The API is designed for the world of blocks and takes WordPress history of being closely attached to web standards to heart.
+
+As directives are added to the HTML, they work great with dynamic blocks and PHP.
```
```suggestion
As directives are HTML attributes, they are perfect for dynamic blocks and PHP.
```

cbravobernal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+
+### Optional and gradual adoption
+
+The Interactivity API pipeline promotes **progressive enhancement** by building on top of WordPress’s solid foundation and patterns. It was carefully designed not to force any use cases to pay for the costs of other use cases.
```
```suggestion
The Interactivity API pipeline promotes **progressive enhancement** by building on top of WordPress’s solid foundation and patterns.
```

cbravobernal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+
+The Interactivity API pipeline promotes **progressive enhancement** by building on top of WordPress’s solid foundation and patterns. It was carefully designed not to force any use cases to pay for the costs of other use cases.
+
+For example, blocks with directives can coexist with other (interactive or non-interactive) blocks. This means that if there are other blocks on the page using other frameworks like jQuery, everything will work as expected.
```
Except full client side navigation. We can add the link to the "supports.interactivity" section.

juanmaguitar replied:
I have added a few [callouts](https://developer.wordpress.org/block-editor/contributors/documentation/#callout-notices) to warn about this. Let me know what you think.

cbravobernal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+
+## Why a standard?
+
+Blocks using the Interactivity API and interactive blocks using other approaches like jQuery can coexist, and everything will work as expected. However, the Interactivity API comes with some benefits for your interactive blocks:
```
Again, beware of client side navigation.

juanmaguitar replied:
See https://github.com/WordPress/gutenberg/pull/61323#discussion_r1590755874

cbravobernal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+
+Blocks using the Interactivity API and interactive blocks using other approaches like jQuery can coexist, and everything will work as expected. However, the Interactivity API comes with some benefits for your interactive blocks:
+
+- **Blocks can communicate with each other easily**. With a standard, this communication is handled by default. When different blocks use different approaches to frontend interactivity, inter-block communication becomes more complex and almost impossible when separate developers create blocks.
```
```suggestion
- **Blocks can communicate with each other easily**. With a standard, this communication is handled by default. When different blocks use different approaches to frontend interactivity, inter-block communication becomes more complex and almost impossible when different developers create blocks.
```

juanmaguitar committed:
Can the Interactivity API be used beyond a block

SantosGuillamot commented on docs/reference-guides/interactivity-api/iapi-faq.md:
```diff
@@ -0,0 +1,129 @@
+
+## Can I use directives in the block editor?
+
+No. Right now, directives only work in the frontend of blocks. However, it’ll be investigated whether some directives (and also your custom directives) could be reused across the frontend and the editor. It’s worth emphasizing that the realities of the editor application and the frontend of a website are very different, particularly around the interactivity they afford. It needs to be ensured tha ...
```
> However, it’ll be investigated whether some directives (and also your custom directives) could be reused across the frontend and the editor.

Other folks could confirm but this doesn't look like a priority right now. Maybe it is explored at some point, but I am a bit afraid that mentioning it could cause expectations.

juanmaguitar replied:
I have removed this question (and answer) to avoid false expectations. If we see this is a recurrent question from the community, we can think of a better response (hopefully providing specific plans) and include it here.

juanmaguitar committed:
Update docs/reference-guides/interactivity-api/iapi-about.md

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

juanmaguitar committed:
Adjustments based on feedback on the PR

juanmaguitar committed:
merge changes

juanmaguitar committed:
Update docs/reference-guides/interactivity-api/iapi-about.md

Co-authored-by: Carlos Bravo <37012961+cbravobernal@users.noreply.github.com>

juanmaguitar committed:
Removed use directives in the block editor

juanmaguitar committed:
Merge branch 'docs/interactivity-api-what-and-faq' of github.com:WordPress/gutenberg into docs/interactivity-api-what-and-faq

sirreal commented on docs/reference-guides/interactivity-api/iapi-faq.md:
```diff
@@ -0,0 +1,124 @@
+
+## Can the Interactivity API be used beyond a block?
+
+Absolutely, yes, it is not limited to blocks. You'll see a lot of mentions of how the Interactivity API provides a standard for creating interactive blocks, but that's only because that's the most common use case. More generally speaking, the Interactivity API standard can be used to add "interactive behaviors" to the front end of any part of WordPress.
```
👍 Good, thanks!

I'm not sure if we want to limit this so we're not getting into implementation, but we could link to [`wp_interactivity_process_directives`](https://developer.wordpress.org/reference/functions/wp_interactivity_process_directives/).

> See the `wp_interactivity_process_directives` for details on using the Interactivity API outside of blocks with arbitrary HTML.

juanmaguitar replied:
I have added this reference to `wp_interactivity_process_directives`

juanmaguitar committed:
Translation API with the iAPI

juanmaguitar committed:
Update docs/reference-guides/interactivity-api/iapi-about.md

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

juanmaguitar committed:
replaced scripts by script modules

juanmaguitar committed:
Update docs/reference-guides/interactivity-api/iapi-about.md

Co-authored-by: Carlos Bravo <37012961+cbravobernal@users.noreply.github.com>

juanmaguitar committed:
Update docs/reference-guides/interactivity-api/iapi-about.md

Co-authored-by: Carlos Bravo <37012961+cbravobernal@users.noreply.github.com>

juanmaguitar committed:
Added wp_interactivity_process_directives to Can the Interactivity API be used beyond a block

juanmaguitar committed:
Merge branch 'docs/interactivity-api-what-and-faq' of github.com:WordPress/gutenberg into docs/interactivity-api-what-and-faq

sirreal commented on docs/reference-guides/interactivity-api/iapi-about.md:
```diff
@@ -0,0 +1,191 @@
+ 
+<div
+  data-wp-interactive='wpmovies'
+  data-wp-context="{ 'isOpen': true }"
```
```suggestion
  <?php echo wp_interactivity_data_wp_context( array( 'isOpen' => true ) ); ?>
```

juanmaguitar committed:
Callouts about full site editing

sirreal commented on docs/reference-guides/interactivity-api/iapi-faq.md:
```diff
@@ -0,0 +1,128 @@
+
+Absolutely, yes, it is not limited to blocks. You'll see a lot of mentions of how the Interactivity API provides a standard for creating interactive blocks, but that's only because that's the most common use case. More generally speaking, the Interactivity API standard can be used to add "interactive behaviors" to the front end of any part of WordPress.
+
+See the [`wp_interactivity_process_directives`](https://developer.wordpress.org/reference/functions/wp_interactivity_process_directives/) for details on using the Interactivity API outside of blocks with arbitrary HTML.
```
```suggestion
See the [`wp_interactivity_process_directives` function](https://developer.wordpress.org/reference/functions/wp_interactivity_process_directives/) for details on using the Interactivity API outside of blocks with arbitrary HTML.
```

juanmaguitar committed:
Update docs/reference-guides/interactivity-api/iapi-faq.md

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

juanmaguitar commented:
@sirreal @cbravobernal @SantosGuillamot I think I have addressed all your comments and feedback in the new commits for this PR. Let me know your thoughts on this last version.

juanmaguitar committed:
Update docs/reference-guides/interactivity-api/iapi-about.md

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

juanmaguitar committed:
Update docs/reference-guides/interactivity-api/iapi-about.md

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

sirreal reviewed (approved):
This seems good to me when you're satisfied. Thanks!

juanmaguitar committed:
better phrasing for client-side navigation exception

juanmaguitar committed:
Merge branch 'docs/interactivity-api-what-and-faq' of github.com:WordPress/gutenberg into docs/interactivity-api-what-and-faq

juanmaguitar committed:
added link to ongoing explorations for scripts loading delay
